Pos files can not be executed on the robot

Getting an error in lola node output on the robot while trying to execute the pos file on a robot.
The same file works fine in simulation.
```
[ERROR] [1692590320.768159872] [msgpack packer]: Incorrect message received for nao_command_msgs::msg::JointStiffnesses. Stiffnesses and Indexes vector must have the same length. Stiffnesses vector has length 0, while indexes vector has length 27
[ros2run]: Segmentation fault
```

--- a/src/indexes.hpp
+++ b/src/indexes.hpp
@@ -1,0 +1,54 @@
+// Copyright 2023 Kenji Brameld
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef INDEXES_HPP_
+#define INDEXES_HPP_
+
+#include <vector>
+
+#include "nao_lola_command_msgs/msg/joint_indexes.hpp"
+
+namespace indexes
+{
+
+static const std::vector<uint8_t> indexes = {
+  nao_lola_command_msgs::msg::JointIndexes::HEADYAW,
+  nao_lola_command_msgs::msg::JointIndexes::HEADPITCH,
+  nao_lola_command_msgs::msg::JointIndexes::LSHOULDERPITCH,
+  nao_lola_command_msgs::msg::JointIndexes::LSHOULDERROLL,
+  nao_lola_command_msgs::msg::JointIndexes::LELBOWYAW,
+  nao_lola_command_msgs::msg::JointIndexes::LELBOWROLL,
+  nao_lola_command_msgs::msg::JointIndexes::LWRISTYAW,
+  nao_lola_command_msgs::msg::JointIndexes::LHIPYAWPITCH,
+  nao_lola_command_msgs::msg::JointIndexes::LHIPROLL,
+  nao_lola_command_msgs::msg::JointIndexes::LHIPPITCH,
+  nao_lola_command_msgs::msg::JointIndexes::LKNEEPITCH,
+  nao_lola_command_msgs::msg::JointIndexes::LANKLEPITCH,
+  nao_lola_command_msgs::msg::JointIndexes::LANKLEROLL,
+  nao_lola_command_msgs::msg::JointIndexes::RHIPROLL,
+  nao_lola_command_msgs::msg::JointIndexes::RHIPPITCH,
+  nao_lola_command_msgs::msg::JointIndexes::RKNEEPITCH,
+  nao_lola_command_msgs::msg::JointIndexes::RANKLEPITCH,
+  nao_lola_command_msgs::msg::JointIndexes::RANKLEROLL,
+  nao_lola_command_msgs::msg::JointIndexes::RSHOULDERPITCH,
+  nao_lola_command_msgs::msg::JointIndexes::RSHOULDERROLL,
+  nao_lola_command_msgs::msg::JointIndexes::RELBOWYAW,
+  nao_lola_command_msgs::msg::JointIndexes::RELBOWROLL,
+  nao_lola_command_msgs::msg::JointIndexes::RWRISTYAW,
+  nao_lola_command_msgs::msg::JointIndexes::LHAND,
+  nao_lola_command_msgs::msg::JointIndexes::RHAND};
+
+}  // namespace indexes
+
+#endif  // INDEXES_HPP_

--- a/src/naosoccer_pos_action_node.cpp
+++ b/src/naosoccer_pos_action_node.cpp
@@ -24,39 +24,13 @@
 #include "boost/filesystem.hpp"
 #include "rclcpp/rclcpp.hpp"
 
+#include "indexes.hpp"
 #include "parser.hpp"
 
 namespace fs = boost::filesystem;
 
 namespace naosoccer_pos_action
 {
-
-std::vector<uint8_t> indexes = {
-  nao_lola_command_msgs::msg::JointIndexes::HEADYAW,
-  nao_lola_command_msgs::msg::JointIndexes::HEADPITCH,
-  nao_lola_command_msgs::msg::JointIndexes::LSHOULDERPITCH,
-  nao_lola_command_msgs::msg::JointIndexes::LSHOULDERROLL,
-  nao_lola_command_msgs::msg::JointIndexes::LELBOWYAW,
-  nao_lola_command_msgs::msg::JointIndexes::LELBOWROLL,
-  nao_lola_command_msgs::msg::JointIndexes::LWRISTYAW,
-  nao_lola_command_msgs::msg::JointIndexes::LHIPYAWPITCH,
-  nao_lola_command_msgs::msg::JointIndexes::LHIPROLL,
-  nao_lola_command_msgs::msg::JointIndexes::LHIPPITCH,
-  nao_lola_command_msgs::msg::JointIndexes::LKNEEPITCH,
-  nao_lola_command_msgs::msg::JointIndexes::LANKLEPITCH,
-  nao_lola_command_msgs::msg::JointIndexes::LANKLEROLL,
-  nao_lola_command_msgs::msg::JointIndexes::RHIPROLL,
-  nao_lola_command_msgs::msg::JointIndexes::RHIPPITCH,
-  nao_lola_command_msgs::msg::JointIndexes::RKNEEPITCH,
-  nao_lola_command_msgs::msg::JointIndexes::RANKLEPITCH,
-  nao_lola_command_msgs::msg::JointIndexes::RANKLEROLL,
-  nao_lola_command_msgs::msg::JointIndexes::RSHOULDERPITCH,
-  nao_lola_command_msgs::msg::JointIndexes::RSHOULDERROLL,
-  nao_lola_command_msgs::msg::JointIndexes::RELBOWYAW,
-  nao_lola_command_msgs::msg::JointIndexes::RELBOWROLL,
-  nao_lola_command_msgs::msg::JointIndexes::RWRISTYAW,
-  nao_lola_command_msgs::msg::JointIndexes::LHAND,
-  nao_lola_command_msgs::msg::JointIndexes::RHAND};
 
 NaosoccerPosActionNode::NaosoccerPosActionNode(const rclcpp::NodeOptions & options)
 : rclcpp::Node{"NaosoccerPosActionNode", options}
@@ -147,7 +121,7 @@ void NaosoccerPosActionNode::calculateEffectorJoints(
 
   if (firstTickSinceActionStarted) {
     nao_lola_command_msgs::msg::JointPositions command;
-    command.indexes = indexes;
+    command.indexes = indexes::indexes;
     command.positions = std::vector<float>(
       sensor_joints.positions.begin(), sensor_joints.positions.end());
     keyFrameStart =
@@ -177,7 +151,7 @@ void NaosoccerPosActionNode::calculateEffectorJoints(
       beta)).c_str());
 
   nao_lola_command_msgs::msg::JointPositions effector_joints;
-  effector_joints.indexes = indexes;
+  effector_joints.indexes = indexes::indexes;
 
   for (unsigned int i = 0; i < nao_lola_command_msgs::msg::JointIndexes::NUMJOINTS; ++i) {
     float previous = previousKeyFrame.positions.positions.at(i);

--- a/src/parser.cpp
+++ b/src/parser.cpp
@@ -20,6 +20,7 @@
 
 #include "rclcpp/logging.hpp"
 #include "nao_lola_command_msgs/msg/joint_indexes.hpp"
+#include "indexes.hpp"
 
 // +2 because there is the "!" at the start, and the duration at the end
 #define POSITIONS_SIZE (nao_lola_command_msgs::msg::JointIndexes::NUMJOINTS + 2)
@@ -27,7 +28,8 @@
 #define STIFFNESSES_SIZE (nao_lola_command_msgs::msg::JointIndexes::NUMJOINTS + 1)
 
 const auto stiffnessMax = nao_lola_command_msgs::msg::JointStiffnesses()
-  .set__indexes({1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1});
+  .set__indexes(indexes::indexes)
+  .set__stiffnesses({1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1});
 
 namespace parser
 {
@@ -61,6 +63,7 @@ ParseResult parse(const std::vector<std::string> & in)
 
       // Convert to data type
       nao_lola_command_msgs::msg::JointPositions jointPositions;
+      jointPositions.indexes = indexes::indexes;
       for (unsigned int i = 1; i < nao_lola_command_msgs::msg::JointIndexes::NUMJOINTS + 1; ++i) {
         std::string position_deg_string = splitted_line.at(i);
 
@@ -111,6 +114,7 @@ ParseResult parse(const std::vector<std::string> & in)
       }
 
       // Convert to data type
+      jointStiffnesses.stiffnesses.clear();
       for (unsigned int i = 1; i < nao_lola_command_msgs::msg::JointIndexes::NUMJOINTS + 1; ++i) {
         std::string stiffness_string = splitted_line.at(i);
 

--- a/test/test_parser.cpp
+++ b/test/test_parser.cpp
@@ -15,7 +15,10 @@
 #include <cmath>
 
 #include "gtest/gtest.h"
+#include "nao_lola_command_msgs/msg/joint_indexes.hpp"
 #include "../src/parser.hpp"
+
+using JointIndexes = nao_lola_command_msgs::msg::JointIndexes;
 
 TEST(TestParser, TestIncorrectPositionSize)
 {
@@ -27,6 +30,83 @@ TEST(TestParser, TestIncorrectPositionSize)
   EXPECT_FALSE(parseResult.successful);
 }
 
+TEST(TestParser, TestPositionsAndStiffnessesSize)
+{
+  std::vector<std::string> testString = {
+    "! 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 300",
+  };
+
+  auto parseResult = parser::parse(testString);
+  EXPECT_EQ(parseResult.keyFrames.at(0).stiffnesses.stiffnesses.size(), JointIndexes::NUMJOINTS);
+  EXPECT_EQ(parseResult.keyFrames.at(0).positions.positions.size(), JointIndexes::NUMJOINTS);
+}
+
+TEST(TestParser, TestIndexesFilled)
+{
+  std::vector<std::string> testString = {
+    "! 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 300",
+  };
+
+  auto parseResult = parser::parse(testString);
+  const auto & key_frame = parseResult.keyFrames.at(0);
+
+  // Position indexes
+  ASSERT_EQ(key_frame.positions.indexes.size(), JointIndexes::NUMJOINTS);
+  EXPECT_EQ(key_frame.positions.indexes.at(0), JointIndexes::HEADYAW);
+  EXPECT_EQ(key_frame.positions.indexes.at(1), JointIndexes::HEADPITCH);
+  EXPECT_EQ(key_frame.positions.indexes.at(2), JointIndexes::LSHOULDERPITCH);
+  EXPECT_EQ(key_frame.positions.indexes.at(3), JointIndexes::LSHOULDERROLL);
+  EXPECT_EQ(key_frame.positions.indexes.at(4), JointIndexes::LELBOWYAW);
+  EXPECT_EQ(key_frame.positions.indexes.at(5), JointIndexes::LELBOWROLL);
+  EXPECT_EQ(key_frame.positions.indexes.at(6), JointIndexes::LWRISTYAW);
+  EXPECT_EQ(key_frame.positions.indexes.at(7), JointIndexes::LHIPYAWPITCH);
+  EXPECT_EQ(key_frame.positions.indexes.at(8), JointIndexes::LHIPROLL);
+  EXPECT_EQ(key_frame.positions.indexes.at(9), JointIndexes::LHIPPITCH);
+  EXPECT_EQ(key_frame.positions.indexes.at(10), JointIndexes::LKNEEPITCH);
+  EXPECT_EQ(key_frame.positions.indexes.at(11), JointIndexes::LANKLEPITCH);
+  EXPECT_EQ(key_frame.positions.indexes.at(12), JointIndexes::LANKLEROLL);
+  EXPECT_EQ(key_frame.positions.indexes.at(13), JointIndexes::RHIPROLL);
+  EXPECT_EQ(key_frame.positions.indexes.at(14), JointIndexes::RHIPPITCH);
+  EXPECT_EQ(key_frame.positions.indexes.at(15), JointIndexes::RKNEEPITCH);
+  EXPECT_EQ(key_frame.positions.indexes.at(16), JointIndexes::RANKLEPITCH);
+  EXPECT_EQ(key_frame.positions.indexes.at(17), JointIndexes::RANKLEROLL);
+  EXPECT_EQ(key_frame.positions.indexes.at(18), JointIndexes::RSHOULDERPITCH);
+  EXPECT_EQ(key_frame.positions.indexes.at(19), JointIndexes::RSHOULDERROLL);
+  EXPECT_EQ(key_frame.positions.indexes.at(20), JointIndexes::RELBOWYAW);
+  EXPECT_EQ(key_frame.positions.indexes.at(21), JointIndexes::RELBOWROLL);
+  EXPECT_EQ(key_frame.positions.indexes.at(22), JointIndexes::RWRISTYAW);
+  EXPECT_EQ(key_frame.positions.indexes.at(23), JointIndexes::LHAND);
+  EXPECT_EQ(key_frame.positions.indexes.at(24), JointIndexes::RHAND);
+
+  // Stiffness indexes
+  ASSERT_EQ(key_frame.stiffnesses.indexes.size(), JointIndexes::NUMJOINTS);
+  EXPECT_EQ(key_frame.stiffnesses.indexes.at(0), JointIndexes::HEADYAW);
+  EXPECT_EQ(key_frame.stiffnesses.indexes.at(1), JointIndexes::HEADPITCH);
+  EXPECT_EQ(key_frame.stiffnesses.indexes.at(2), JointIndexes::LSHOULDERPITCH);
+  EXPECT_EQ(key_frame.stiffnesses.indexes.at(3), JointIndexes::LSHOULDERROLL);
+  EXPECT_EQ(key_frame.stiffnesses.indexes.at(4), JointIndexes::LELBOWYAW);
+  EXPECT_EQ(key_frame.stiffnesses.indexes.at(5), JointIndexes::LELBOWROLL);
+  EXPECT_EQ(key_frame.stiffnesses.indexes.at(6), JointIndexes::LWRISTYAW);
+  EXPECT_EQ(key_frame.stiffnesses.indexes.at(7), JointIndexes::LHIPYAWPITCH);
+  EXPECT_EQ(key_frame.stiffnesses.indexes.at(8), JointIndexes::LHIPROLL);
+  EXPECT_EQ(key_frame.stiffnesses.indexes.at(9), JointIndexes::LHIPPITCH);
+  EXPECT_EQ(key_frame.stiffnesses.indexes.at(10), JointIndexes::LKNEEPITCH);
+  EXPECT_EQ(key_frame.stiffnesses.indexes.at(11), JointIndexes::LANKLEPITCH);
+  EXPECT_EQ(key_frame.stiffnesses.indexes.at(12), JointIndexes::LANKLEROLL);
+  EXPECT_EQ(key_frame.stiffnesses.indexes.at(13), JointIndexes::RHIPROLL);
+  EXPECT_EQ(key_frame.stiffnesses.indexes.at(14), JointIndexes::RHIPPITCH);
+  EXPECT_EQ(key_frame.stiffnesses.indexes.at(15), JointIndexes::RKNEEPITCH);
+  EXPECT_EQ(key_frame.stiffnesses.indexes.at(16), JointIndexes::RANKLEPITCH);
+  EXPECT_EQ(key_frame.stiffnesses.indexes.at(17), JointIndexes::RANKLEROLL);
+  EXPECT_EQ(key_frame.stiffnesses.indexes.at(18), JointIndexes::RSHOULDERPITCH);
+  EXPECT_EQ(key_frame.stiffnesses.indexes.at(19), JointIndexes::RSHOULDERROLL);
+  EXPECT_EQ(key_frame.stiffnesses.indexes.at(20), JointIndexes::RELBOWYAW);
+  EXPECT_EQ(key_frame.stiffnesses.indexes.at(21), JointIndexes::RELBOWROLL);
+  EXPECT_EQ(key_frame.stiffnesses.indexes.at(22), JointIndexes::RWRISTYAW);
+  EXPECT_EQ(key_frame.stiffnesses.indexes.at(23), JointIndexes::LHAND);
+  EXPECT_EQ(key_frame.stiffnesses.indexes.at(24), JointIndexes::RHAND);
+}
+
 TEST(TestParser, TestParsePosition)
 {
   std::vector<std::string> testString = {
@@ -35,9 +115,23 @@ TEST(TestParser, TestParsePosition)
 
   auto parseResult = parser::parse(testString);
   ASSERT_TRUE(parseResult.successful);
-  ASSERT_EQ(parseResult.keyFrames.size(), 1);
-  EXPECT_EQ(parseResult.keyFrames.at(0).t_ms, 300);
+  ASSERT_EQ(parseResult.keyFrames.size(), 1u);
+  EXPECT_EQ(parseResult.keyFrames.at(0).t_ms, 300u);
   EXPECT_NEAR(parseResult.keyFrames.at(0).positions.positions.at(1), 90 * M_PI / 180.0, 0.0001);
+}
+
+TEST(TestParser, TestStiffnessIsOneByDefault)
+{
+  std::vector<std::string> testString = {
+    "! 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 300",
+  };
+
+  auto parseResult = parser::parse(testString);
+  const auto & keyFrameOneStiffnesses = parseResult.keyFrames.at(0).stiffnesses.stiffnesses;
+  ASSERT_EQ(keyFrameOneStiffnesses.size(), JointIndexes::NUMJOINTS);
+  for (unsigned i = 0; i < JointIndexes::NUMJOINTS; ++i) {
+    EXPECT_EQ(keyFrameOneStiffnesses.at(i), 1.0);
+  }
 }
 
 TEST(TestParser, TestParseStiffness)
@@ -49,8 +143,8 @@ TEST(TestParser, TestParseStiffness)
 
   auto parseResult = parser::parse(testString);
   ASSERT_TRUE(parseResult.successful);
-  ASSERT_EQ(parseResult.keyFrames.size(), 1);
-  EXPECT_EQ(parseResult.keyFrames.at(0).t_ms, 300);
+  ASSERT_EQ(parseResult.keyFrames.size(), 1u);
+  EXPECT_EQ(parseResult.keyFrames.at(0).t_ms, 300u);
   EXPECT_NEAR(parseResult.keyFrames.at(0).stiffnesses.stiffnesses.at(1), 0.2, 0.0001);
 }
 
@@ -65,7 +159,7 @@ TEST(TestParser, TestDuration)
 
   auto parseResult = parser::parse(testString);
   ASSERT_TRUE(parseResult.successful);
-  ASSERT_EQ(parseResult.keyFrames.size(), 2);
-  EXPECT_EQ(parseResult.keyFrames.at(0).t_ms, 300);
-  EXPECT_EQ(parseResult.keyFrames.at(1).t_ms, 600);
+  ASSERT_EQ(parseResult.keyFrames.size(), 2u);
+  EXPECT_EQ(parseResult.keyFrames.at(0).t_ms, 300u);
+  EXPECT_EQ(parseResult.keyFrames.at(1).t_ms, 600u);
 }

